File: worker/agents/notebookRetention.ts
import type Redis from "ioredis";
import { makeBus } from "../lib/bus";
import {
  NotebookLMClient,
  notebookLMConfig,
  retention,
  type ArticleNotebookState,
  type NotebookKind,
} from "../../src/lib/core/notebooklm";

export interface NotebookCleanupJob {
  uid?: string;
  jobId?: string;
}

const NOTEBOOK_KINDS: NotebookKind[] = ["articles", "courses"];

/** Evict lowest-engagement (tiebreak soonest-expiry) sources over the cap. */
async function evictOverflow(redis: Redis, client: NotebookLMClient, notebook: NotebookKind): Promise<number> {
  const cfg = client.cfg;
  const notebookId = cfg.notebooks[notebook];
  let count = await retention.sourceCount(redis, notebookId);
  if (count <= cfg.sourceCap) return 0;

  const candidates: ArticleNotebookState[] = [];
  for (const id of await retention.activeArticleIds(redis)) {
    const st = await retention.getArticleState(redis, id);
    if (st && st.active && st.notebook === notebook && st.sourceId) candidates.push(st);
  }
  candidates.sort((a, b) => a.score - b.score || a.expiresAt - b.expiresAt);

  let evicted = 0;
  for (const st of candidates) {
    if (count <= cfg.sourceCap) break;
    try {
      await client.removeSource(notebook, st.sourceId!);
    } catch (e) {
      console.warn("[notebookCleanup] overflow remove failed:", (e as Error).message);
    }
    await retention.markRemoved(redis, st.articleId, notebookId);
    count--;
    evicted++;
  }
  return evicted;
}

export interface NotebookEngagementJob {
  uid?: string;
  jobId?: string;
  articleId: string;
  /** 0..100 engagement/quiz score. Defaults to a mid "viewed" weight. */
  score?: number;
}

/**
 * Rotation/cleanup job: remove sources whose retention window has passed from the
 * Articles notebook, advancing each to "removed". Keeps the notebook under
 * NotebookLM's per-notebook source cap. Safe to run on a timer.
 */
export async function runNotebookCleanup(redis: Redis, job: NotebookCleanupJob = {}): Promise<{ removed: number }> {
  const cfg = notebookLMConfig();
  if (!cfg.enabled) {
    if (job.uid && job.jobId) {
      await makeBus(redis).publish(job.uid, { jobId: job.jobId, type: "done", status: "ready", result: { removed: 0, skipped: true } });
    }
    return { removed: 0 };
  }

  const client = new NotebookLMClient(cfg);
  const due = await retention.dueForRemoval(redis, Date.now());
  let removed = 0;

  for (const articleId of due) {
    const st = await retention.getArticleState(redis, articleId);
    if (st?.sourceId) {
      try {
        await client.removeSource(st.notebook, st.sourceId);
      } catch (e) {
        console.warn("[notebookCleanup] remove failed:", (e as Error).message);
      }
    }
    const notebookId = st ? cfg.notebooks[st.notebook] : undefined;
    await retention.markRemoved(redis, articleId, notebookId);
    removed++;
  }

  // Then keep each notebook under its source cap (lowest-engagement evicted first).
  let evicted = 0;
  for (const notebook of NOTEBOOK_KINDS) {
    evicted += await evictOverflow(redis, client, notebook);
  }
  removed += evicted;

  if (job.uid && job.jobId) {
    await makeBus(redis).publish(job.uid, { jobId: job.jobId, type: "done", status: "ready", result: { removed, evicted } });
  }
  return { removed };
}

/**
 * Engagement signal: a view/listen/good quiz score extends an article's retention
 * window (so engaging material survives the next rotation).
 */
export async function runRecordEngagement(redis: Redis, job: NotebookEngagementJob): Promise<void> {
  const cfg = notebookLMConfig();
  const state = await retention.touchEngagement(redis, job.articleId, job.score ?? 50, cfg.retentionDays);
  if (job.uid && job.jobId) {
    await makeBus(redis).publish(job.uid, {
      jobId: job.jobId,
      type: "done",
      status: "ready",
      result: { articleId: job.articleId, status: state?.status ?? "unknown", expiresAt: state?.expiresAt },
    });
  }
}

File: worker_py/agents/notebook_retention.py
"""Retention/rotation + engagement agents. Port of ``worker/agents/notebookRetention.ts``."""

from typing import Dict, Optional

from redis.asyncio import Redis

from ..bus import publish as bus_publish
from ..notebooklm import NotebookLMClient, notebooklm_config, retention
from ..util import now_ms

_NOTEBOOK_KINDS = ["articles", "courses"]


async def _evict_overflow(redis: Redis, client: NotebookLMClient, notebook: str) -> int:
    """Evict lowest-engagement (tiebreak soonest-expiry) sources over the cap."""
    cfg = client.cfg
    notebook_id = cfg.notebooks[notebook]
    count = await retention.source_count(redis, notebook_id)
    if count <= cfg.source_cap:
        return 0

    candidates = []
    for article_id in await retention.active_article_ids(redis):
        st = await retention.get_article_state(redis, article_id)
        if st and st.get("active") and st.get("notebook") == notebook and st.get("sourceId"):
            candidates.append(st)
    candidates.sort(key=lambda s: (s.get("score", 0), s.get("expiresAt", 0)))

    evicted = 0
    for st in candidates:
        if count <= cfg.source_cap:
            break
        try:
            await client.remove_source(notebook, st["sourceId"])
        except Exception as e:  # noqa: BLE001
            print(f"[notebook_cleanup] overflow remove failed: {e}")
        await retention.mark_removed(redis, st["articleId"], notebook_id)
        count -= 1
        evicted += 1
    return evicted


async def run_notebook_cleanup(redis: Redis, job: Optional[Dict] = None) -> Dict[str, int]:
    """Remove sources whose retention window has passed; keep the notebook under
    NotebookLM's per-notebook source cap. Safe to run on a timer."""
    job = job or {}
    cfg = notebooklm_config()
    if not cfg.enabled:
        if job.get("uid") and job.get("jobId"):
            await bus_publish(redis, job["uid"], {"jobId": job["jobId"], "type": "done", "status": "ready", "result": {"removed": 0, "skipped": True}})
        return {"removed": 0}

    client = NotebookLMClient(cfg)
    due = await retention.due_for_removal(redis, now_ms())
    removed = 0

    for article_id in due:
        st = await retention.get_article_state(redis, article_id)
        if st and st.get("sourceId"):
            try:
                await client.remove_source(st["notebook"], st["sourceId"])
            except Exception as e:  # noqa: BLE001
                print(f"[notebook_cleanup] remove failed: {e}")
        notebook_id = cfg.notebooks[st["notebook"]] if st else None
        await retention.mark_removed(redis, article_id, notebook_id)
        removed += 1

    # Then keep each notebook under its source cap (lowest-engagement evicted first).
    evicted = 0
    for notebook in _NOTEBOOK_KINDS:
        evicted += await _evict_overflow(redis, client, notebook)
    removed += evicted

    if job.get("uid") and job.get("jobId"):
        await bus_publish(redis, job["uid"], {"jobId": job["jobId"], "type": "done", "status": "ready", "result": {"removed": removed, "evicted": evicted}})
    return {"removed": removed}


async def run_record_engagement(redis: Redis, job: Dict) -> None:
    """A view/listen/good quiz score extends an article's retention window."""
    cfg = notebooklm_config()
    state = await retention.touch_engagement(redis, job["articleId"], job.get("score") or 50, cfg.retention_days)
    if job.get("uid") and job.get("jobId"):
        await bus_publish(
            redis,
            job["uid"],
            {
                "jobId": job["jobId"],
                "type": "done",
                "status": "ready",
                "result": {"articleId": job["articleId"], "status": (state or {}).get("status", "unknown"), "expiresAt": (state or {}).get("expiresAt")},
            },
        )
